tools/autograd/templates/variable_factories.h

```diff
@@ -186,12 +186,23 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
 /// support it in the future by iterating over all sub-lists to find
 /// the largest data type that can represent all of the elements, or by using
 /// variadic templates.
-inline at::Tensor tensor(detail::InitListTensor list_init_tensor, const at::TensorOptions& options) {
-  return autograd::make_variable(list_init_tensor.to_tensor(options), options.requires_grad());
+inline at::Tensor tensor(detail::InitListTensor init_list_tensor, const at::TensorOptions& options) {
+  return autograd::make_variable(init_list_tensor.to_tensor(options), options.requires_grad());
 }
 
-inline at::Tensor tensor(detail::InitListTensor list_init_tensor) {
-  return torch::tensor(list_init_tensor, at::dtype(list_init_tensor.scalar_type()));
+inline at::Tensor tensor(detail::InitListTensor init_list_tensor) {
+  return torch::tensor(init_list_tensor, at::dtype(init_list_tensor.scalar_type()));
+}
+
+/// NOTE: We add `torch::tensor(std::initializer_list<detail::InitListTensor>)` function overload,
+/// so that `torch::tensor({{1, 2}})` can take this overload instead of `torch::tensor(at::ArrayRef<T>)`.
+inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list) {
+  TORCH_CHECK(
+    init_list.begin()->type() != detail::InitListTensorType::Scalar,
+    "1D tensor construction such as `torch::tensor({1, 2, 3})` should never take the ",
+    "torch::tensor(std::initializer_list<detail::InitListTensor>) function overload. ",
+    "Please fix the code to avoid this regression.")
```

Comment:
What happens if one does `tensor({1, 1.1})`? Does that still give a reasonable error message?

Reply:
Currently `tensor({1, 1.1})` gives compile error:
```
../test/cpp/api/tensor.cpp: In member function ‘virtual void TensorTest_ContainsCorrectValuesForManyValuesVariable_Test::TestBody()’:
../test/cpp/api/tensor.cpp:230:38: error: call of overloaded ‘tensor(<brace-enclosed initializer list>)’ is ambiguous
   std::cout << torch::tensor({1, 1.1}) << std::endl;
                                      ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:181:644: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:181:766: note: candidate: at::Tensor torch::tensor(std::initializer_list<unsigned char>)
../torch/csrc/autograd/generated/variable_factories.h:181:1603: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:181:1724: note: candidate: at::Tensor torch::tensor(std::initializer_list<signed char>)
../torch/csrc/autograd/generated/variable_factories.h:181:2562: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<short int>)
../torch/csrc/autograd/generated/variable_factories.h:181:2685: note: candidate: at::Tensor torch::tensor(std::initializer_list<short int>)
../torch/csrc/autograd/generated/variable_factories.h:181:3507: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<int>)
../torch/csrc/autograd/generated/variable_factories.h:181:3624: note: candidate: at::Tensor torch::tensor(std::initializer_list<int>)
../torch/csrc/autograd/generated/variable_factories.h:181:4450: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<long int>)
../torch/csrc/autograd/generated/variable_factories.h:181:4572: note: candidate: at::Tensor torch::tensor(std::initializer_list<long int>)
../torch/csrc/autograd/generated/variable_factories.h:181:5404: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<float>)
../torch/csrc/autograd/generated/variable_factories.h:181:5525: note: candidate: at::Tensor torch::tensor(std::initializer_list<float>)
../torch/csrc/autograd/generated/variable_factories.h:181:6354: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<double>)
../torch/csrc/autograd/generated/variable_factories.h:181:6477: note: candidate: at::Tensor torch::tensor(std::initializer_list<double>)
../torch/csrc/autograd/generated/variable_factories.h:181:7630: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<bool>)
../torch/csrc/autograd/generated/variable_factories.h:181:7815: note: candidate: at::Tensor torch::tensor(std::initializer_list<bool>)
../torch/csrc/autograd/generated/variable_factories.h:181:9224: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:181:9409: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::Half>)
../torch/csrc/autograd/generated/variable_factories.h:181:10838: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<c10::BFloat16>)
../torch/csrc/autograd/generated/variable_factories.h:181:11031: note: candidate: at::Tensor torch::tensor(std::initializer_list<c10::BFloat16>)
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:197:19: note: candidate: at::Tensor torch::tensor(torch::detail::InitListTensor)
 inline at::Tensor tensor(detail::InitListTensor init_list_tensor) {
                   ^
In file included from ../caffe2/../torch/csrc/api/include/torch/types.h:7:0,
                 from ../caffe2/../torch/csrc/api/include/torch/detail/static.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/pimpl.h:4,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/module.h:3,
                 from ../caffe2/../torch/csrc/api/include/torch/nn/cloneable.h:3,
                 from ../test/cpp/api/support.h:7,
                 from ../test/cpp/api/tensor.cpp:2:
../torch/csrc/autograd/generated/variable_factories.h:212:19: note: candidate: at::Tensor torch::tensor(std::initializer_list<torch::detail::InitListTensor>)
 inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list) {
                   ^
```
And it doesn't work on v1.2 either:
```
/data/libtorch/test/example-app.cpp:4:34: error: call of overloaded ‘tensor(<brace-enclosed initializer list>)’ is ambiguous
   auto t = torch::tensor({1, 1.2});
                                  ^
In file included from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/types.h:7:0,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader_options.h:4,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader/base.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader/stateful.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data/dataloader.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/data.h:3,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/all.h:4,
                 from /data/libtorch/libtorch/include/torch/csrc/api/include/torch/torch.h:3,
                 from /data/libtorch/test/example-app.cpp:1:
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:644: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<unsigned char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:766: note: candidate: at::Tensor torch::tensor(std::initializer_list<unsigned char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:1603: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<signed char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:1724: note: candidate: at::Tensor torch::tensor(std::initializer_list<signed char>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:2562: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<short int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:2685: note: candidate: at::Tensor torch::tensor(std::initializer_list<short int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:3507: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:3624: note: candidate: at::Tensor torch::tensor(std::initializer_list<int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:4450: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<long int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:4572: note: candidate: at::Tensor torch::tensor(std::initializer_list<long int>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:5404: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<float>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:5525: note: candidate: at::Tensor torch::tensor(std::initializer_list<float>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:6354: note: candidate: at::Tensor torch::tensor(c10::ArrayRef<double>)
/data/libtorch/libtorch/include/torch/csrc/autograd/generated/variable_factories.h:47:6477: note: candidate: at::Tensor torch::tensor(std::initializer_list<double>)
```
So this is not a regression, and it's consistent with our message that `torch::tensor(...)` with mixed types doesn't work at the moment.